pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml

```diff
@@ -1474,8 +1474,281 @@ public class Foo {
+        <code><![CDATA[
+import java.time.LocalDate;
+public record Person(String name, LocalDate birthDate) {}
+        ]]></code>
+    </test-code>
+
+    <!-- Test cases for imports used in sealed classes -->
+    <test-code>
+        <description>Import used in sealed class hierarchy</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import some.pkg.Shape;
+public sealed class Circle extends Shape permits BigCircle, SmallCircle {}
+        ]]></code>
+    </test-code>
+
+    <!-- Test cases for imports used in pattern matching -->
+    <test-code>
+        <description>Import used in instanceof pattern matching</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import java.util.List;
+public class Foo {
+    void test(Object o) {
+        if (o instanceof List l) {
+            l.size();
+        }
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <!-- Test cases for imports used in text blocks -->
+    <test-code>
+        <description>Import used in text block with embedded expressions</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import java.time.LocalDate;
+public class Foo {
+    String test(String name) {
+        return """
+            Hello %s!
+            Today is %s
+            """.formatted(name, LocalDate.now());
+    }
 }
         ]]></code>
     </test-code>
 
+    <!-- Test cases for imports used in var declarations -->
+    <test-code>
+        <description>Import used in var declaration with initializer</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import java.util.ArrayList;
+public class Foo {
+    void test() {
+        var list = new ArrayList<String>();
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <!-- Test cases for imports used in module-info.java -->
+<!--    <test-code>-->
+<!--        <description>Import used in module-info requires</description>-->
+<!--        <expected-problems>0</expected-problems>-->
+<!--        <code><![CDATA[-->
+<!--import some.module.ModuleName;-->
+<!--module my.module {-->
+<!--    requires ModuleName;-->
+<!--}-->
+<!--        ]]></code>-->
+<!--    </test-code>-->
+
+    <!-- Test cases for imports used in Java 14+ record patterns -->
+    <test-code>
+        <description>Import used in record pattern</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import some.pkg.Point;
+public class Foo {
+    void test(Object o) {
+        if (o instanceof Point(int x, int y)) {
+            System.out.println(x + "," + y);
+        }
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <!-- Test cases for imports used in Java 22+ unnamed patterns/variables -->
+    <test-code>
+        <description>Import used with unnamed pattern variable</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import some.pkg.Point;
```

Comment:
This test case should be improved: Please create the related classes/types in src/test/java in package `net.sourceforge.pmd.lang.java.rule.codestyle.unnecessaryimport` so that we actually use type resolution correctly.